rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,189 @@
+  type into Vector will not be considered.
+
+## Pain
+
+## Proposal
+
+### User Experience
+
+This adds a new sink to Vector to send data into Google Chronicle.
+
+There is a choice of two different types of events that can be sent to Chronicle.
+
+Which type to send will be configured with an `event_type` configuration option.
+
+```toml
+event_type = "unstructured"
+```
+
+or
+
+```toml
+event_type = "udm"
+```
+
+#### UDM (Unified Data Model) events
+
+https://cloud.google.com/chronicle/docs/reference/ingestion-api#udmevents
+https://cloud.google.com/chronicle/docs/unified-data-model/udm-usage
+
+UDM events are events that conform to a [strict schema](https://cloud.google.com/chronicle/docs/reference/udm-field-list).
+An example event:
+
+```json
+{
+  "metadata": {
+    "event_timestamp": "2019-10-23T04:00:00.000Z",
+    "event_type": "NETWORK_HTTP",
+    "product_name": "Acme Proxy",
+    "vendor_name": "Acme"
+  },
+  "network": {
+    "http": {
+      "method": "GET",
+      "response_code": 200
+    }
+  },
+  "principal": {
+    "hostname": "host0",
+    "ip": "10.1.2.3",
+    "port": 60000
+  },
+  "target": {
+    "hostname": "www.altostrat.com",
+    "ip": "198.51.100.68",
+    "port": 443,
+    "url": "www.altostrat.com/images/logo.png"
+  }
+}
+```
+
+Events can be batched and sent in a JSON object that also includes the
+`customer_id`:
+
+```json
+{
+"customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+"events": [...]
+}
+```
+
+UDM events can be sumbitted as either JSON or Proto3. We will submit using JSON
+as this fits in easier with Vectors data model. The newly developed schema
+support for Vector will be used to ensure the JSON is formed according to the
+[Chronicle requirements](https://cloud.google.com/chronicle/docs/unified-data-model/format-events-as-udm).
+
+Vector has no internal knowledge of SIEM events so it will be up to the user to
+ensure the event is well formed - either the source will receive the data in the
+correct structure, or it will be transformed into such before reaching the sink.
+
+#### Unstructured log data events
+
+Unstructured log events are events in plain text with no strict schema attached.
+
+```json
+{
+  "log_text": "26-Feb-2019 13:35:02.187 client 10.120.20.32#4238: query: altostrat.com IN A + (203.0.113.102)",
+  "ts_epoch_microseconds": 1551188102187000
+}
+```
+
+Events are batched and sent in a JSON object that includes the `customer_id` and
+a `log_type`:
+
+```json
+{
+  "customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+  "log_type": "BIND_DNS",
+  "entries": [...]
+}
+```
+
+`log_type` has to be one of the values returned by the [logtypes endpoint](https://cloud.google.com/chronicle/docs/reference/ingestion-api#logtypes).
+
+If `event_type == "unstructured"` we will require another configuration field -
+`log_type` to be set. This field will be templatable.
+
+Another field `remove_log_type` allows the field used in the template to be
+removed from the event.
```

Comment:
Alternatively, we add support for "enum variants" in schema requirements, which would allow you to require `log_type` to be one of a list of strings. We can then treat this field as any other: use a `remap` transform to add the field, if Vector fails to boot at startup, because of the missing field (or an invalid string variant being supplied).

Reply:
I'm not sure at this stage whether this is a fixed list or if users are able to customise this through the Chronicle UI, but we can get the list via an API call: https://cloud.google.com/chronicle/docs/reference/ingestion-api#logtypes

Theoretically, we call this at boot time and use the list to validate this field at run time. I don't think it is worth doing that since Chronicle itself will validate this. I have added it as an outstanding question.